packages/block-library/src/social-link/edit.js

```diff
@@ -25,39 +25,53 @@ import {
 } from '@wordpress/components';
 import { __, sprintf } from '@wordpress/i18n';
 import { keyboardReturn } from '@wordpress/icons';
+import { isURL, isEmail } from '@wordpress/url';
 
 /**
  * Internal dependencies
  */
-import { getIconBySite, getNameBySite } from './social-list';
+import {
+	getIconBySite,
+	getNameBySite,
+	getMatchingService,
+} from './social-list';
 
 const SocialLinkURLPopover = ( {
 	url,
 	setAttributes,
-	setPopover,
+	onClose,
 	popoverAnchor,
 	clientId,
 } ) => {
 	const { removeBlock } = useDispatch( blockEditorStore );
 	return (
-		<URLPopover
-			anchor={ popoverAnchor }
-			onClose={ () => setPopover( false ) }
-		>
+		<URLPopover anchor={ popoverAnchor } onClose={ () => onClose( false ) }>
 			<form
 				className="block-editor-url-popover__link-editor"
 				onSubmit={ ( event ) => {
 					event.preventDefault();
-					setPopover( false );
+					onClose( false );
 				} }
 			>
 				<div className="block-editor-url-input">
 					<URLInput
 						__nextHasNoMarginBottom
 						value={ url }
-						onChange={ ( nextURL ) =>
-							setAttributes( { url: nextURL } )
-						}
+						onChange={ ( nextURL ) => {
+							const nextAttributes = { url: nextURL };
+
+							if ( isURL( nextURL ) || isEmail( nextURL ) ) {
+								const matchingService = isEmail( nextURL )
+									? 'mail'
+									: getMatchingService( nextURL );
+
+								if ( matchingService ) {
+									nextAttributes.service = matchingService;
+								}
+							}
+
+							setAttributes( nextAttributes );
+						} }
```

Comment:
Maybe we should provide a hit below the URL input for this new feature. Not sure about the design or text thought 😅